dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/user/UserController.java

```diff
@@ -517,45 +495,57 @@ public void putJsonObject( @PathVariable( "uid" ) String pvUid, HttpServletReque
 
-        if ( !userService.canAddOrUpdateUser( IdentifiableObjectUtils.getUids( parsed.getGroups() ), currentUser )
+        if ( !userService.canAddOrUpdateUser( groupsUids, currentUser )
             || !currentUser.getUserCredentials().canModifyUser( users.get( 0 ).getUserCredentials() ) )
         {
-            throw new WebMessageException( WebMessageUtils.conflict( "You must have permissions to create user, or ability to manage at least one user group for the user." ) );
+            throw new WebMessageException( WebMessageUtils.conflict(
+                "You must have permissions to create user, " +
+                    "or ability to manage at least one user group for the user." ) );
         }
 
         MetadataImportParams params = importService.getParamsFromMap( contextService.getParameterValuesMap() );
         params.setImportReportMode( ImportReportMode.FULL );
         params.setImportStrategy( ImportStrategy.UPDATE );
-        params.addObject( parsed );
+        params.addObject( parsedUserObject );
 
         ImportReport importReport = importService.importMetadata( params );
 
-        updateUserGroups( importReport, pvUid, parsed, currentUser );
-
-        response.setContentType( "application/json" );
-
-        renderService.toJson( response.getOutputStream(), importReport );
-    }
-
-    protected void updateUserGroups( ImportReport importReport, String pvUid, User parsed, User currentUser )
-    {
         if ( importReport.getStatus() == Status.OK && importReport.getStats().getUpdated() == 1 )
         {
-            User user = userService.getUser( pvUid );
+            updateUserGroups( userUid, parsedUserObject, currentUser );
 
-            // current user may have been changed and detached and must become managed again
-            if ( currentUser != null && currentUser.getId() == user.getId() )
+            // If it was a pw change attempt (input.pw != null) and update was success we assume password has changed...
+            // We chose to expire the special case if password is set to the same. i.e. no before & after equals pw check
+            if ( isPasswordChangeAttempt )
             {
-                currentUser = currentUserService.getCurrentUser();
+                userService.expireActiveSessions( parsedUserObject.getUserCredentials() );
             }
+        }
 
-            userGroupService.updateUserGroups( user, IdentifiableObjectUtils.getUids( parsed.getGroups() ), currentUser );
+        return importReport;
+    }
+
+    protected void updateUserGroups( String pvUid, User parsed, User currentUser )
+    {
+        User user = userService.getUser( pvUid );
+
+        // current user may have been changed and detached and must become managed again
+        // TODO: what is this doing? I don't understand how this is possible.
+        if ( currentUser != null && currentUser.getId() == user.getId() )
```

Comment:
Here again related to currentUser == null. But here with a twist currentUser == input.user then reset current user, can someone please explain how this will help?